stgit/commands/common.py

```diff
@@ -461,6 +463,19 @@ def run_commit_msg_hook(repo, cd, editor_is_used=True):
         return cd
 
 
+def run_pre_commit_hook(repo):
+    pre_commit_hook = get_hook(repo, 'pre-commit')
```

Comment:
I am not sure if we have to change the `GIT_INDEX_FILE` variable by ourselves. Apparently GIT sets it to a temp index while running hooks and locks the original one as per this stackoverflow answer - https://stackoverflow.com/questions/9437063/how-to-add-a-file-to-the-index-in-a-git-pre-commit-hook

Reply:
This is a better explanation of the way git handles `pre-commit` hook in `git commit`.

https://stackoverflow.com/questions/56524769/why-do-pre-commit-hooks-leave-files-half-staged

Look for `More than three copies` in the first answer.

Reply:
One more relevant link: https://stackoverflow.com/questions/51587518/whats-the-difference-between-git-commit-file-and-git-commit-only/51587817

Reply:
In order to set `GIT_INDEX_FILE` for the `pre-commit` script, we would need to call the hook in `write_tree()` and in the context of the temporary `iw` object. This further requires modifications to `get_hook()` to take an `iw` parameter instead of a `repository` parameter.

I've experimented with this idea. The code for `write_tree()` looks something like this:

```python
def write_tree(stack, paths, temp_index, no_verify):
    """Possibly update the index, and then write its tree. If any path
    limiting is in effect, use a temp index.
    @return: The written tree.
    @rtype: L{Tree<stgit.git.Tree>}"""

    def go(index):
        if paths:
            iw = IndexAndWorktree(index, stack.repository.default_worktree)
            iw.update_index(paths)
            pre_commit_hook = get_hook(iw, 'pre-commit')
            if not no_verify and pre_commit_hook:
                try:
                    pre_commit_hook()
                except RunException:
                    raise CmdException(
                        'pre-commit hook failed, review the changes using `stg diff`, '
                        'run `stg add` to add them to index and run `stg refresh` again'
                    )
                index.read_tree(stack.head)
                iw.update_index(paths)
        return index.write_tree()

    if temp_index:
        index = stack.repository.temp_index()
        try:
            index.read_tree(stack.head)
            return go(index)
        finally:
            index.delete()
            stack.repository.default_iw.update_index(paths)
    else:
        return go(stack.repository.default_index)
```

A key thing to take note of is that `IndexAndWorktree.run()` injects `GIT_INDEX_FILE` into the environment. So by supplying the temp `iw` instance to `get_hook()`, it will can run the `pre-commit` script with the temp `iw` instead of its current behavior of using `default_iw`.

Per [my other comment](https://github.com/stacked-git/stgit/pull/63#discussion_r463985880), I do not believe that we must make this change, but since the experiment you describe above failed to use a non-default `GIT_INDEX_FILE`, I thought it might be helpful to illustrate the problem with that experiment.

Reply:
Thanks Pete. This was very helpful.

When I first started working on this feature, this was my first thought to break up the `write_tree` function in 2 parts and return index from first part, use that to run hook and write the actually tree only afterwards. Because otherwise the solution is suboptimal. Though I could not come up with a satisfactory solution but your implementation makes sense now.